ADBDEV-5039: Do not treat interruption of LDAP requests by signals as an auth error.

RekGRpth opened the pull request:
Do not treat interruption of LDAP requests by signals as an auth error.

When authorizing via ldap, the ldap_simple_bind_s function may return an error.
This occurs when an internal call to the ldap_int_poll utility function is
interrupted by a signal. The openldap library can handle this properly using the
LDAP_OPT_RESTART option. Setting this option to LDAP_OPT_ON causes the select()
system call to be restarted when interrupted by a signal. The default value is
LDAP_OPT_OFF. This patch sets the value to LDAP_OPT_ON (also it requires
explicitly specifying the -llber library to build) after ldap is initialized.

Tests are not provided because they require a separate LDAP container and the
error is not reproducible in a stable manner.

RekGRpth committed:
Do not treat interruption of LDAP requests by signals as an auth error.

When authorizing via ldap, the ldap_simple_bind_s function may return an error.
This occurs when an internal call to the ldap_int_poll utility function is
interrupted by a signal. The openldap library can handle this properly using the
LDAP_OPT_RESTART option. Setting this option to LDAP_OPT_ON causes the select()
system call to be restarted when interrupted by a signal. The default value is
LDAP_OPT_OFF. This patch sets the value to LDAP_OPT_ON (also it requires
explicitly specifying the -llber library to build) after ldap is initialized.

Tests are not provided because they require a separate LDAP container and the
error is not reproducible in a stable manner.

BenderArenadata commented:
Failed job Build for x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1199933

BenderArenadata commented:
Failed job Build for ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1199934

RekGRpth commented:
bender build

BenderArenadata commented:
Failed job Build for ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1200021

BenderArenadata commented:
Failed job Build for x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1200020

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/67349

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1200179

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1200172

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1200180

whitehawk commented:
Summary 'Interruption of LDAP requests by signals shouldn't be treated as auth error' looks violating wiki checklist for patch description, item 2.2 (summary should be in Imperative Mood).

Following wiki I guess it should be smth like "Do not treat interruption of LDAP requests by signals as auth error" or similar.

whitehawk commented:
In the PR description there is no explanation why no tests were added. Refer to wiki checklist for patch description, item 3.1.

RekGRpth commented:
> Summary 'Interruption of LDAP requests by signals shouldn't be treated as auth error' looks violating wiki checklist for patch description, item 2.2 (summary should be in Imperative Mood).
> 
> Following wiki I guess it should be smth like "Do not treat interruption of LDAP requests by signals as auth error" or similar.

updated



> In the PR description there is no explanation why no tests were added. Refer to wiki checklist for patch description, item 3.1.

added

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/68656

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1261006

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1261007

andr-sokolov commented on src/backend/libpq/auth.c:
```diff
@@ -2554,6 +2554,7 @@ static int
 InitializeLDAPConnection(Port *port, LDAP **ldap)
 {
 	int			ldapversion = LDAP_VERSION3;
+	int			ldaprestart = 1;
```
According to https://linux.die.net/man/3/ldap_set_option for LDAP_OPT_RESTART "value should either be LDAP_OPT_OFF or LDAP_OPT_ON". I suggest to use LDAP_OPT_ON instead of 1

RekGRpth replied:
applied

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/68728

RekGRpth committed:
optimize

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/68736

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1267217

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1267218

Commit:
Merge branch 'adb-6.x-dev' into ADBDEV-5039

swork12 commented:
bender build

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/69397

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1302730
